tox.ini

```diff
@@ -1,42 +1,27 @@
 [tox]
 min_version = 4.0
-env_list = format, test_pydantic1, type_pydantic1, test_pydantic2, type_pydantic2, lint, py{3.9,3.10,3.11}
+env_list = format, lint, py{39,310,311}-pydantic{1,2}
 
 [gh-actions]
 python =
     3.9: py39
     3.10: py310
-    3.11: py311, test_pydantic1, type_pydantic1, test_pydantic2, type_pydantic2, format, lint
+    3.11: format, lint, py311
 
-[testenv:test_pydantic1]
+[testenv]
 labels = core
 allowlist_externals = poetry
+# The "pydanticX:" prefixes are Tox factor-conditional settings.
+# https://tox.wiki/en/3.4.0/config.html?highlight=conditional#factors-and-factor-conditional-settings
+# Note that "poetry add" changes pyproject.toml, but at least we
+# change it back when the tests finish.
 commands_pre =
-  poetry install --only main --only test --no-root --all-extras
-  poetry add pydantic^1
-commands = poetry run pytest -vv tests
-
-[testenv:type_pydantic1]
-allowlist_externals = poetry
-commands_pre =
-  poetry install --no-root --all-extras
-  poetry add pydantic^1
-commands = poetry run mypy openapi_pydantic tests
-
-[testenv:test_pydantic2]
-labels = core
-allowlist_externals = poetry
-commands_pre =
-  poetry install --only main --only test --no-root --all-extras
-  poetry add 'pydantic>=1.8'
-commands = poetry run pytest -vv tests
-
-[testenv:type_pydantic2]
-allowlist_externals = poetry
-commands_pre =
+  pydantic1: poetry add pydantic<2
+  pydantic2: poetry add pydantic>=1.8
   poetry install --no-root --all-extras
-  poetry add 'pydantic>=1.8'
-commands = poetry run mypy openapi_pydantic tests
+commands =
+  poetry run pytest -vv tests
+  poetry run mypy openapi_pydantic tests
```

Comment:
nit: I'd prefer if type checking stays in a seperate env to tests to keep output cleaner?

Reply:
Ok, I figured out how to run the unit tests and type checking in separate jobs while still testing with multiple versions of Python and Pydantic.